Support multidimensional coordinates for bounds and regridding

tomvothecoder opened the pull request:
## Description
<!--
  Please include a summary of the change and which issue is fixed.
  Please also include relevant motivation and context.
  List any dependencies that are required for this change.
-->

- Closes #816

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

If applicable:

- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass with my changes (locally and CI/CD build)
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have noted that this is a breaking change for a major release (fix or feature that would cause existing functionality to not work as expected)

tomvothecoder committed:
Initial code to support multidimensional coordinates

tomvothecoder committed:
Remove unused code

tomvothecoder committed:
Remove unused code

tomvothecoder committed:
Apply suggestion from @tomvothecoder

tomvothecoder commented:
Notes
- Add guard for regrid2 since it not supported

pochedls commented:
Thanks for pushing this forward @jasonb5! This PR seems to work for the example in #816, but I think it is failing for a related case (curvilinear->rectilinear):

```bash
wget https://esgf-node.ornl.gov/thredds/fileServer/css03_data/CMIP6/CMIP/CNRM-CERFACS/CNRM-CM6-1/historical/r1i1p1f2/Amon/tas/gr/v20180917/tas_Amon_CNRM-CM6-1_historical_r1i1p1f2_gr_185001-201412.nc
wget https://esgf-node.ornl.gov/thredds/fileServer/css03_data/CMIP6/CMIP/CNRM-CERFACS/CNRM-CM6-1/historical/r1i1p1f2/SImon/siconc/gn/v20180917/siconc_SImon_CNRM-CM6-1_historical_r1i1p1f2_gn_185001-201412.nc
```
```python
# %% imports
import xcdat as xc

# %% parameters
fn_source = 'siconc_SImon_CNRM-CM6-1_historical_r1i1p1f2_gn_185001-201412.nc'
fn_target = 'tas_Amon_CNRM-CM6-1_historical_r1i1p1f2_gr_185001-201412.nc'

# %% get source / target grids
ds = xc.open_dataset(fn_source)
ngrid = xc.open_dataset(fn_target)

# %% try regridding with xcdat
dsr = ds.regridder.horizontal('tmp2m', ngrid, tool="xesmf", method="conservative_normed", periodic=True)
```

> KeyError: "No 'X' axis dimension coordinate variables were found in the xarray object. Make sure dimension coordinate variables exist, they are one dimensional, and their CF 'axis' or 'standard_name' attrs are correctly set."

I can get this to work with:

```python
import xesmf as xe

regridder = xe.Regridder(ds, ngrid, "conservative_normed", periodic=True, ignore_degenerate=True)                                            
dsr = regridder(ds)
```

jasonb5 commented:
@pochedls Could you give this another try? Thanks!

pochedls commented:
This is working for both of the cases I raised (code below for my reference), but I think it isn't generically picking up the coordinates. 

For example, [CESM2](https://esgf-node.ornl.gov/thredds/fileServer/css03_data/CMIP6/CMIP/NCAR/CESM2/historical/r1i1p1f1/SImon/siconc/gn/v20190308/siconc_SImon_CESM2_historical_r1i1p1f1_gn_185001-201412.nc) has the following structure

> Dimensions:    (time: 1980, nj: 384, ni: 320, d2: 2, nvertices: 4, bnds: 2)
> Coordinates:
>   * time       (time) object 16kB 1850-01-15 12:00:00 ... 2014-12-15 12:00:00
>   * nj         (nj) int32 2kB 1 2 3 4 5 6 7 8 ... 378 379 380 381 382 383 384
>   * ni         (ni) int32 1kB 1 2 3 4 5 6 7 8 ... 314 315 316 317 318 319 320
>     lat        (nj, ni) float64 983kB dask.array<chunksize=(384, 320), meta=np.ndarray>
>     lon        (nj, ni) float64 983kB dask.array<chunksize=(384, 320), meta=np.ndarray>
> Dimensions without coordinates: d2, nvertices, bnds
> Data variables:
>     siconc     (time, nj, ni) float32 973MB dask.array<chunksize=(1, 384, 320), meta=np.ndarray>

Which produces the following error when regridding:

> ValueError: This DataArray has more than one dimension ['lon', 'nj'] mapped to the 'X' axis, which is an unexpected behavior. Try dropping extraneous dimensions from the DataArray first (might affect data shape).

Is there a more generic way to handle these auxiliary coordinates (x/nj/etc)?

Perlmutter code for reference.

```python
import xsearch as xs
import xcdat as xc

p = list(xs.findPaths('historical', 'siconc', 'mon', model='CNRM-CM6-1', member = 'r1i1p1f2').keys())[0]
p2 = list(xs.findPaths('historical', 'tas', 'mon', model='CNRM-CM6-1', member = 'r1i1p1f2').keys())[0]

ds = xc.open_mfdataset(p)
ngrid = xc.open_mfdataset(p2)
dsr = ds.regridder.horizontal('siconc', ngrid, tool="xesmf", method="conservative_normed", periodic=True)
```

jasonb5 commented:
That actually caught an edge case I didn't consider. If the `multidim` flag is set it should return a matched coordinate no matter what. I'll push a fix and have you retest. Thanks!

pochedls commented:
> That actually caught an edge case I didn't consider. If the `multidim` flag is set it should return a matched coordinate no matter what. I'll push a fix and have you retest. Thanks!

Thanks @jasonb5 – I think curvilinear regridding is seemingly full of edge cases. I appreciate your effort in making this robust (and I'm happy you can quickly understand what is going on).

jasonb5 commented:
@pochedls There's more to this issue. My original fix may not be correct. I'm working on a new fix and will push it soon.

jasonb5 committed:
Adds support for multidimensional coordinates in regridders

jasonb5 commented:
@pochedls Try the latest. The CESM example will fail because it has two possible `X` coordinates `('nj', 'lon')` so xCDAT can differentiate. It looks like `cf_xarray` can figure out what are the correct coordinates. I may switch `ds.coords` to `ds.cf.coordinates` to fix this. Will update after some testing.

pochedls commented:
> @pochedls Try the latest. The CESM example will fail because it has two possible `X` coordinates `('nj', 'lon')` so xCDAT can differentiate. It looks like `cf_xarray` can figure out what are the correct coordinates. I may switch `ds.coords` to `ds.cf.coordinates` to fix this. Will update after some testing.

FYI – this is working for CNRM-CM6-1 and the `source.nc` data in the original bug report. It is failing for CESM2 (as you expected). Once you update for CESM2, I can try a broader set of models

jasonb5 committed:
Fix mypy error

jasonb5 committed:
Fixes where coordinates are sourced when allowing multi-dim coords

jasonb5 commented:
@pochedls Go ahead and try again. All three cases should work this time.

pochedls commented:
@jasonb5 – This is working for all the examples I provided. I'm good with moving forward with this PR (fixes the issues raised), but I could also test more models if that would be helpful.

jasonb5 commented:
@pochedls Thanks!
@tomvothecoder Do you want to review before we merge?

tomvothecoder commented:
Thanks @jasonb5 for the work and @pochedls for testing. I will review and merge early next week.

tomvothecoder commented on xcdat/regridder/accessor.py:
```diff
@@ -304,14 +304,20 @@ def _obj_to_grid_ds(obj: xr.Dataset | xr.DataArray) -> xr.Dataset:
+
     for axis, has_bounds in axis_has_bounds.items():
         if not has_bounds:
+            # FIXME: Line 313 --Can't add bounds for multidimensional coordinates
```
Is this `#FIXME` still relevant or should it be a `#TODO`? Also exact line number should be removed.

jasonb5 replied:
No longer relevant, will remove.

tomvothecoder commented on xcdat/regridder/base.py:
```diff
@@ -95,6 +95,12 @@ def _drop_axis(ds: xr.Dataset, axis: list[CFAxisKey]) -> xr.Dataset:
+    @classmethod
+    def can_handle_multidim(cls) -> bool:
+        return cls.supports_multidim
+
```
Good point

jasonb5 replied:
Fixed, defaulting to False.

tomvothecoder commented on xcdat/axis.py:
```diff
@@ -117,10 +118,14 @@ def get_dim_coords(
     ----------
     .. [1] https://cf-xarray.readthedocs.io/en/latest/coord_axes.html#axes-and-coordinates
     """
-    # Get the object's index keys, with each being a dimension.
-    # NOTE: xarray does not include multidimensional coordinates as index keys.
-    # Example: ["lat", "lon", "time"]
-    index_keys = obj.indexes.keys()
+    if multidim:
+        # multidimensional coordinates cannot be indexes, use all coords
+        index_keys = list([y for x in obj.cf.coordinates.values() for y in x])
+    else:
+        # Get the object's index keys, with each being a dimension.
+        # NOTE: xarray does not include multidimensional coordinates as index keys.
+        # Example: ["lat", "lon", "time"]
+        index_keys = list(obj.indexes.keys())
```
Good point, this is more robust than only using `obj.cf.coordinates`.

jasonb5 replied:
Combined `cf.coordinates` and `cf.axes`, should have complete coverage.

tomvothecoder commented on xcdat/regridder/accessor.py:
```diff
@@ -166,7 +167,9 @@ def horizontal(
             ) from e
 
-        input_grid = _get_input_grid(self._ds, data_var, ["X", "Y"])
+        input_grid = _get_input_grid(
+            self._ds, data_var, ["X", "Y"], multidim=regrid_tool.can_handle_multidim()
+        )
```
Test needed.

jasonb5 replied:
Added missing test.

tomvothecoder commented on xcdat/regridder/accessor.py:
```diff
@@ -304,14 +304,20 @@ def _obj_to_grid_ds(obj: xr.Dataset | xr.DataArray) -> xr.Dataset:
         attrs=obj.attrs,
     )
 
+    # Multidimensional coordinates bounds generation is not supported
+    if multidim:
+        return output_ds
```
I think Copilot is referencing the logic before this conditional in `_obj_to_grid_ds()`:

https://github.com/xCDAT/xcdat/blob/d13a72ab2c09f00c2b704163d018922aa41a8f7f/xcdat/regridder/accessor.py#L281-L305

tomvothecoder commented:
@jasonb5 I tagged Copilot for review and it caught some things. Can you review its suggestions and decide how and where to take action? Thanks.

jasonb5 committed:
Removes old comment

jasonb5 committed:
Set default supports_multidim to False on BaseRegridder

Set a default value for supports_multidim so subclasses are not required
to explicitly declare it. Added tests to verify the default and override
behavior.

jasonb5 committed:
Include CF axes in multidimensional coordinate discovery in get_dim_coords

The multidim=True branch only used obj.cf.coordinates to build
index_keys, missing coordinates discoverable only via obj.cf.axes.
Combine both sources to avoid incorrect KeyError when an axis
coordinate exists but is not in cf.coordinates.

jasonb5 committed:
Add regression test for #816 multidimensional coordinate regridding

Exercises ds.regridder.horizontal(..., tool='xesmf') on a dataset with
2D lat/lon coordinates and no 1D index coord vars, ensuring the
multidim code path correctly discovers coordinates via CF axis metadata.

jasonb5 committed:
Pass multidimensional flag through _get_axis_coord_and_bounds

Datasets with non-standard 2D coordinate names (e.g., nav_lat/nav_lon)
that are only CF-discoverable via axis metadata were failing because
_get_axis_coord_and_bounds called get_dim_coords without multidim=True.
Pass the flag through so the full coord discovery path works.

jasonb5 commented:
@pochedls @tomvothecoder Made the changes, might one to do one more pass before we merge.

pochedls commented:
This is failing for the CESM2 case again (works for CNRM-CM6-1 and the original example issue). 

```                                                                                                                                       
import xcdat as xc

p = '/global/cfs/projectdirs/m4931/gsharing/css03_data/CMIP6/CMIP/NCAR/CESM2/historical/r1i1p1f1/SImon/siconc/gn/v20190308/' 
p2 = '/global/cfs/projectdirs/m4931/gsharing/css03_data/CMIP6/CMIP/NCAR/CESM2/historical/r1i1p1f1/Amon/tas/gn/v20190308/'

ds = xc.open_mfdataset(p)
ngrid = xc.open_mfdataset(p2)
dsr = ds.regridder.horizontal('siconc', ngrid, tool="xesmf", method="conservative_normed", periodic=True)
```

> ValueError: This DataArray has more than one dimension ['lon', 'nj'] mapped to the 'X' axis, which is an unexpected behavior. Try dropping extraneous dimensions from the DataArray first (might affect data shape).

tomvothecoder commented:
@jasonb5 will push fix and @pochedls to test it. Uses coordinates to fall-back to axis if axis check doesn't work. Uses cf-xarray to do that.

jasonb5 committed:
Use cf coordinates then fallback to axes when processing multi dimensional coordinates

jasonb5 commented:
@pochedls Try this again, should resolve the use cases. I'll need to address this error in another PR.

tomvothecoder commented:
@jasonb5 I rebased this branch on the latest main, which includes a fix to a breaking xgcm API change (#840). The unit tests in the GH Actions build should pass now.

jasonb5 committed:
Merge branch 'main' into bug/816-multidim-coords

pochedls commented:
@jasonb5 – thanks for your work on this! This appears to be fixed for all of the cases I raised.

tomvothecoder commented:
@jasonb5 Checking to see if you have time for final review and merge

tomvothecoder commented:
Thanks for taking care of this @jasonb5
